File: examples/ruby-openai/test_client.rb
# frozen_string_literal: true
require "minitest/autorun"
require "webrick"
require_relative "main"

class ClientTest < Minitest::Test
  def test_typed_requests_do_not_retry_errors
    [200, 429, 503].each do |status|
      calls = []
      server = WEBrick::HTTPServer.new(Port: 0, BindAddress: "127.0.0.1", Logger: WEBrick::Log.new(File::NULL), AccessLog: [])
      server.mount_proc("/") do |request, response|
        calls << [request.path, request["authorization"], JSON.parse(request.body)]
        response.status = status
        response["Content-Type"] = "application/json"
        response.body = JSON.generate(status == 200 ? {id: "example", object: "chat.completion", created: 1, model: "example", choices: []} : {error: {message: "unavailable", type: "provider_unavailable"}})
      end
      thread = Thread.new { server.start }
      http = local_http
      begin
        client = stogas_client("http://127.0.0.1:#{server.config[:Port]}/cap/v1", "test-key", http)
        request = -> { client.chat.completions.create(model: "example", messages: [{role: "user", content: "hello"}]) }
        if status == 200
          assert_equal "example", request.call.id
        else
          assert_raises(OpenAI::Errors::APIStatusError, &request)
        end
        assert_equal 1, calls.length
        assert_equal "/cap/v1/chat/completions", calls[0][0]
        assert_equal "Bearer test-key", calls[0][1]
        assert_equal "example", calls[0][2]["model"]
      ensure
        http.close
        server.shutdown
        thread.join
      end
    end
  end
end

File: examples/python-openai/test_client.py
"""The real OpenAI client against a local transport boundary; no inference costs."""

import json
import threading
import unittest
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from openai import APIStatusError

from main import client


class ClientTests(unittest.TestCase):
    def test_request_objects_errors_and_redirects_never_replay(self):
        for status in (200, 307, 429, 503):
            with self.subTest(status=status):
                calls = []

                class Handler(BaseHTTPRequestHandler):
                    def log_message(self, *args):
                        pass

                    def do_POST(self):
                        calls.append((
                            self.path,
                            self.headers.get("Authorization"),
                            json.loads(self.rfile.read(int(self.headers["Content-Length"]))),
                        ))
                        self.send_response(status)
                        self.send_header("Content-Type", "application/json")
                        self.send_header("Location", "/must-not-follow")
                        self.end_headers()
                        payload = (
                            {"id": "chat-1", "choices": [{"index": 0, "message": {
                                "role": "assistant", "content": "Hello"
                            }, "finish_reason": "stop"}]}
                            if status == 200 else
                            {"error": {"message": "Unavailable", "type": "server_error"}}
                        )
                        self.wfile.write(json.dumps(payload).encode())

                with ThreadingHTTPServer(("127.0.0.1", 0), Handler) as server:
                    thread = threading.Thread(target=server.serve_forever)
                    thread.start()
                    try:
                        with client(f"http://127.0.0.1:{server.server_port}/v1", "example-key") as api:
                            request = {"model": "example/model", "messages": [{"role": "user", "content": "hello"}]}
                            if status == 200:
                                result = api.chat.completions.create(**request)
                                self.assertEqual(result.choices[0].message.content, "Hello")
                            else:
                                with self.assertRaises(APIStatusError) as caught:
                                    api.chat.completions.create(**request)
                                self.assertEqual(caught.exception.status_code, status)
                            self.assertEqual(calls, [("/v1/chat/completions", "Bearer example-key", request)])
                    finally:
                        server.shutdown()
                        thread.join()


if __name__ == "__main__":
    unittest.main()
